cleanrl/sac_continuous_action_jax.py

```diff
@@ -0,0 +1,548 @@
+from dataclasses import dataclass
+from distutils.util import strtobool
+from functools import partial
+from typing import Any, NamedTuple, Optional, Sequence
+
+import flax
+import flax.linen as nn
+import gym
+import jax
+import jax.numpy as jnp
+import numpy as np
+import optax
+import pybullet_envs  # noqa
+import tensorflow_probability
+from flax.training.train_state import TrainState
+from stable_baselines3.common.buffers import ReplayBuffer
+from stable_baselines3.common.env_util import make_vec_env
+from stable_baselines3.common.evaluation import evaluate_policy
+from stable_baselines3.common.vec_env import DummyVecEnv
+from torch.utils.tensorboard import SummaryWriter
+
+# Add progress bar if available
+try:
+    from tqdm.rich import tqdm
+except ImportError:
+    tqdm = None
+
+tfp = tensorflow_probability.substrates.jax
+tfd = tfp.distributions
+
+
+class RLTrainState(TrainState):
+    target_params: flax.core.FrozenDict
+
+
+class ReplayBufferSamplesNp(NamedTuple):
+    observations: np.ndarray
+    actions: np.ndarray
+    next_observations: np.ndarray
+    dones: np.ndarray
+    rewards: np.ndarray
+
+
+class TanhTransformedDistribution(tfd.TransformedDistribution):
+    """
+    From https://github.com/ikostrikov/walk_in_the_park
+    otherwise mode is not defined for Squashed Gaussian
+    """
+
+    def __init__(self, distribution: tfd.Distribution, validate_args: bool = False):
+        super().__init__(distribution=distribution, bijector=tfp.bijectors.Tanh(), validate_args=validate_args)
+
+    def mode(self) -> jnp.ndarray:
+        return self.bijector.forward(self.distribution.mode())
+
+    @classmethod
+    def _parameter_properties(cls, dtype: Optional[Any], num_classes=None):
+        td_properties = super()._parameter_properties(dtype, num_classes=num_classes)
+        del td_properties["bijector"]
+        return td_properties
+
+
+class Critic(nn.Module):
+    n_units: int = 256
+
+    @nn.compact
+    def __call__(self, x: jnp.ndarray, action: jnp.ndarray) -> jnp.ndarray:
+        x = jnp.concatenate([x, action], -1)
+        x = nn.Dense(self.n_units)(x)
+        x = nn.relu(x)
+        x = nn.Dense(self.n_units)(x)
+        x = nn.relu(x)
+        x = nn.Dense(1)(x)
+        return x
+
+
+class VectorCritic(nn.Module):
+    n_units: int = 256
+    n_critics: int = 2
+
+    @nn.compact
+    def __call__(self, obs: jnp.ndarray, action: jnp.ndarray) -> jnp.ndarray:
+        # Idea taken from https://github.com/perrin-isir/xpag
+        # Similar to https://github.com/tinkoff-ai/CORL for PyTorch
+        vmap_critic = nn.vmap(
+            Critic,
+            variable_axes={"params": 0},  # parameters not shared between the critics
+            split_rngs={"params": True},  # different initializations
+            in_axes=None,
+            out_axes=0,
+            axis_size=self.n_critics,
+        )
+        q_values = vmap_critic(
+            n_units=self.n_units,
+        )(obs, action)
+        return q_values
```

Comment:
> This seems a bit over-complicated to me since we usually do 2 critics. 

well, I have mixed feelings.
Using `vmap` shows the true power of jax and also remove a lot of duplicated code (it is even worse when you have dropout enable) and the complexity is contained in the `VectorCritic`.
It is true that it is a bit less explicit than have two critics, two critic states, two random keys. 
This version also allow you to have SAC-N for free (which is good at offline learning apparently).

I will take a second look at it later.